src/osd/pg_scrubber.h

```diff
@@ -0,0 +1,631 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <cassert>
+#include <chrono>
+#include <memory>
+#include <mutex>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <vector>
+
+#include "PG.h"
+#include "ScrubStore.h"
+#include "scrub_machine_lstnr.h"
+#include "scrubber_common.h"
+
+class Callback;
+
+namespace Scrub {
+class ScrubMachine;
+struct BuildMap;
+
+
+/**
+  ReplicaReservations is an RAII wrapper of the process of reserving/freeing scrub
+  resources at the replicas.
+
+  When constructed - sends reservation requests to the acting_set.
+  Handles arriving responses:
+  - a 'grant' is marked by reducing the 'pending_' counter. A list of reserved
+    resources is maintained.
+  - a rejection triggers a "couldn't acquire the replicas' scrub resources" event.
+    All "grants" arriving after the first rejection are denied, and cause an immediate
+    release of the resource.
+
+  Upon destruction - the acting-set (actually - those of the acting-set that have
+  'granted' our request - are freed.
+*/
+class ReplicaReservations {
+  using OrigSet =
+    decltype(std::declval<PG>().get_actingset());  // as I plan to change this type
+  using Vec = std::vector<pg_shard_t>;	// to be replaced with small_vector or small_set
+
+  PG* pg_;  // a pointer, as to not affect ReplicaReservations 'triviality'
+  PgScrubber* scrubber_;
+  OrigSet acting_set_;
+  OSDService* osds_;  // a pointer, as to not affect 'triviality'
+  Vec reserved_peers_;
+  bool had_rejections_{false};
+  int pending_;
```

Comment:
Initializer?

Reply:
The constructor does eventually initialize it, but at least set it to an obviously invalid value so that future constructors that forget to initialize it will at least behave deterministically.

Reply:
Will do that. But as I do not know that value at this point - I am not sure it's a good rule to follow.

Reply:
The value cannot validly be -1, so set it to that.  That way future debug statements have a shot at demonstrating that the member was never initialized properly.  If you leave it this way, the value is nondeterministic which is almost never ok.